// uu_appg01_core-validation/src/matchers/page-info-matcher.js
"use strict";

const Matcher = require("../matcher.js");
const ValidationResult = require("../validation-result.js");
const Validator = require("../validator.js");

const PAGE_INFO_TYPE = `const pageInfoType = shape({
  pageSize: integer(1, null),
  pageIndex: integer(0, null),
  total: integer(-1, null)
}, false);`;

class _PageInfo extends Matcher {

  constructor(...params) {
    super("pageInfo", ...params);
    this.validator = new Validator(PAGE_INFO_TYPE);
  }

  match(value, ctx) {
    if (value == null) {
      return new ValidationResult();
    }

    let result = this.validator.validate("pageInfoType", value);
    if (!result.isValid()) {
      if (result.invalidTypes["shape.e001"]) {
        delete result.invalidTypes["shape.e001"];
        result.addInvalidTypeError("pageInfo.e001", "The value must be pageInfo.");
      }
      if (result.invalidValues["shape.e002"]) {
        delete result.invalidValues["shape.e002"];
        result.addInvalidValueError("pageInfo.e002", "The content of pageInfo must be valid.");
      }
    }

    return result;
  }

}

function PageInfo(...params) {
  let matcher = new _PageInfo(...params);
  if (this && this.matchers && this.matchers[0]) {
    this.matchers.push(matcher);
    return this;
  } else {
    let validator = new Validator(null);
    validator._loadMatchers();
    validator.matchers = [matcher];
    return validator;
  }
}

module.exports = PageInfo;
