// uu_appg01_core-validation/src/matchers/sys-matcher.js
"use strict";

const Matcher = require("../matcher.js");
const ValidationResult = require("../validation-result.js");
const Validator = require("../validator.js");

const SYS_TYPE = `const sysType = shape({
  cts: dateTime().isRequired(),
  mts: dateTime().isRequired(),
  rev: integer().isRequired(),
  lockHash: base64(),
  lockExpTs: datetime()
}, true);`;

class _Sys extends Matcher {

  constructor(...params) {
    super("sys", ...params);
    this.validator = new Validator(SYS_TYPE);
  }

  match(value, ctx) {
    if (value == null) {
      return new ValidationResult();
    }

    let result = this.validator.validate("sysType", value);
    if (!result.isValid()) {
      if (result.invalidTypes["shape.e001"]) {
        delete result.invalidTypes["shape.e001"];
        result.addInvalidTypeError("sys.e001", "The value must be sys.");
      }
      if (result.invalidValues["shape.e002"]) {
        delete result.invalidValues["shape.e002"];
        result.addInvalidValueError("sys.e002", "The content of sys must be valid.");
      }
    }

    return result;
  }

}

function Sys(...params) {
  let matcher = new _Sys(...params);
  if (this && this.matchers && this.matchers[0]) {
    this.matchers.push(matcher);
    return this;
  } else {
    let validator = new Validator(null);
    validator._loadMatchers();
    validator.matchers = [matcher];
    return validator;
  }
}

module.exports = Sys;
