// uu_appg01_devkit/src/config/.eslintrc-uu5.js
const DEFAULT_REACT_VERSION = "16.14.0";
let reactVersion;
try {
  let reactPkg = require("react/package.json", { paths: ["."] });
  reactVersion = reactPkg.version;
} catch (e) {
  reactVersion = DEFAULT_REACT_VERSION;
}

module.exports = {
  env: {
    browser: true,
    commonjs: true,
    es6: true,
    node: true,
    jest: true,
  },
  parser: require.resolve("@babel/eslint-parser"),
  parserOptions: {
    requireConfigFile: false,
    babelOptions: {
      configFile: require.resolve("./.babelrc.js"),
    },
  },
  plugins: ["react", "json", "prettier", "uu5"],
  extends: [
    "eslint:recommended",
    "plugin:react/recommended",
    "prettier",
    "plugin:json/recommended",
    "plugin:uu5/recommended",
  ],
  rules: {
    "no-const-assign": "warn",
    "no-this-before-super": "warn",
    "no-undef": "warn",
    "no-unreachable": "warn",
    "no-unused-vars": ["warn", { args: "none", ignoreRestSiblings: true }],
    "no-use-before-define": ["error", { functions: false }],
    "constructor-super": "warn",
    "valid-typeof": "warn",
    "react/prop-types": "off",
    "react/jsx-closing-bracket-location": "warn",
    "react/jsx-indent-props": ["warn", 2],
    "react/jsx-wrap-multilines": "warn",
    "react/no-direct-mutation-state": "error",
    "react/jsx-no-bind": "off",
    "react/no-deprecated": "off",
    "react/display-name": "off",
    "react/react-in-jsx-scope": "off",
    "no-console": "off",
    "prettier/prettier": "warn",
  },
  settings: {
    react: {
      version: reactVersion,
    },
  },
};
