// uu_appg01_server-client/src/client.browser.js
"use strict";

const {
  AppClient: { AppClient, PerflogHandler, RemoteErrorHandler, SessionHandler, TransformParametersHandler, UriHandler, CsrfTokenHandler }
} = require("uu_appg01_core");
const CurrentSessionHandler = require("./current-session-handler.browser.js");
AppClient.CurrentSessionHandler = CurrentSessionHandler;
const {AccessPolicyVerificationHandler} = require("uu_appg01_oidc").Internal;

const DEFAULT_INTERCEPTORS = [
  PerflogHandler,
  UriHandler,
  ... (AccessPolicyVerificationHandler ? [AccessPolicyVerificationHandler] : []),
  CurrentSessionHandler,
  SessionHandler,
  CsrfTokenHandler,
  RemoteErrorHandler,
  TransformParametersHandler
];

class Client extends AppClient {
  /**
   * Equivalent to {@link UuApp.AppClient.Client#get} using Client with default
   * configuration, i.e.:
   *
   *     new UuApp.AppClient.Client().get(uri, options)
   *
   * @see {@link UuApp.AppClient.Client#get Client#get}
   * @method UuApp.AppClient.Client.get
   */
  static async get(uri, parameters = null, options = null) {
    return new Client().get(uri, parameters, options);
  }

  static async cmdGet(uri, parameters = null, options = null) {
    return new Client().cmdGet(uri, parameters, options);
  }

  /**
   * Equivalent to {@link UuApp.AppClient.Client#post} using Client with default
   * configuration, i.e.:
   *
   *     new UuApp.AppClient.Client().post(uri, parameters, options)
   *
   * @see {@link UuApp.AppClient.Client#post Client#post}
   * @method UuApp.AppClient.Client.post
   */
  static async post(uri, parameters = null, options = null) {
    return new Client().post(uri, parameters, options);
  }

  static async cmdPost(uri, parameters = null, options = null) {
    return new Client().cmdPost(uri, parameters, options);
  }

  static getDefaultInterceptors() {
    return DEFAULT_INTERCEPTORS;
  }
}

module.exports = Client;
