// uu_appg01_devkit-common/src/upgrades/upgrade-5.3.1.js
const path = require("path");

const fs = require("fs-extra");
const klaw = require("klaw");

const upgradeVersion = __filename.replace(/.*upgrade-(.*)\.js$/, "$1");

const DEFAULT_MAPPINGS_JSON = {
  "{vendor}-{uuApp}-{uuSubApp}": {
    useCaseMap: {
      defaultVuc: {
        realization: "index.html",
        httpMethod: "GET",
        type: "VUC",
      },
    },
  },
};

module.exports = async function (config, templateInfo) {
  let { type } = templateInfo;

  let htmlList = [];
  if (type === "uu5-app") {
    htmlList = getUveHtmlFiles();
  }
  if (type.endsWith("lib") || type.startsWith("uu5-")) {
    htmlList = htmlList.concat(await getDemoHtmlFiles());
  }
  if (htmlList.length) {
    console.log(`${upgradeVersion} Adding "lang" attribute to <html> element in UVE and demo *.html files.`);
    for (let htmlFile of htmlList) {
      let content = fs.readFileSync(htmlFile, "utf-8");
      let origContent = content;
      content = content.replace(/<html(\s+[^>]*)?(?=>)/, (m, g) =>
        g && g.match(/\slang\s*=\s*/) ? m : m + ' lang="en"'
      );
      if (content !== origContent) fs.writeFileSync(htmlFile, content, "utf-8");
    }
  }
};

async function getDemoHtmlFiles() {
  let list = [];
  if (fs.existsSync("demo")) {
    for await (const item of klaw("demo")) {
      if (!item.path.endsWith(".html")) continue;
      list.push(item.path);
    }
  }
  return list;
}

function getUveHtmlFiles() {
  let mappingsJson = getMappingsJson();
  let srcPath = "src";

  let resultSet = new Set();
  if (mappingsJson) {
    for (let k in mappingsJson) {
      // iterate over UCs in the mappings.json and pick those which are VUC and end with .html (or .htm)
      let ucMap = (mappingsJson[k] || {})["useCaseMap"] || {};
      Object.keys(ucMap)
        .filter(
          (uc) =>
            (ucMap[uc].type === "VUC" || ucMap[uc].type === "UVE") && (ucMap[uc].realization || "").match(/\.html?$/i)
        )
        .map((uc) => ucMap[uc].realization.replace(/^\/+/, ""))
        .forEach((htmlFile) => resultSet.add(htmlFile));
    }
  }
  let filesInSrc = fs.readdirSync(srcPath, { withFileTypes: true });
  filesInSrc.filter((item) => item.name.match(/\.html?$/) && item.isFile()).forEach((item) => resultSet.add(item.name));
  return [...resultSet].map((it) => path.join(srcPath, it));
}

function getMappingsJson() {
  let serverDirName = path.basename(process.cwd()).replace(/-(client|hi)/, "-server"); // "xyz-hi" => "xyz-server"; "xyz-client" => "xyz-server"

  let filePath = `../${serverDirName}/app/config/mappings.json`;
  if (!fs.existsSync(filePath)) filePath = "config/mappings.json";
  if (!fs.existsSync(filePath)) return DEFAULT_MAPPINGS_JSON;
  let mappingsJson = fs.readFileSync(filePath, "utf-8");
  return mappingsJson ? JSON.parse(mappingsJson) : {};
}
