// uu_appg01_auditlog/src/middleware/auditlog-middleware-start.js
"use strict";
const { DaoFactory } = require("uu_appg01_datastore");
const { Config } = require("uu_appg01_core-utils");
const RequestAuditLogAbl = require("../abl/request-audit-log-abl");

class AuditLogMiddlewareStart {

  constructor() {
    this.name = "uuAuditLog start middleware";
    this.code = "uuAuditLog_start_middleware";
    this.order = 225;
    this.shouldRun = DaoFactory.isDataStoreOn() && Config.getBoolean("uu_app_auditlog_start_middleware");
  }

  async pre(req, res, next) {
    res.locals.auditLogStartTime = Date.now();

    if( this.shouldRun ){
      const requestLogged = await RequestAuditLogAbl.logRequest(this.code, req, res);
      res.locals.auditLogStartHasLogged = !!requestLogged;
    }else{
      res.locals.auditLogStartHasLogged = false;
    }

    return next();
  }

  async ensure(req, res) {
    if( res.locals.auditLogStartHasLogged || !this.shouldRun ){
      return;
    }
    await RequestAuditLogAbl.logRequest(this.code, req, res);
  }

}

module.exports = AuditLogMiddlewareStart;
