// uu_appg01_devkit-common/src/templates-generator/__uu5/visual-component.js
//@@viewOn:imports
import { createVisualComponent, Utils, Content } from "uu5g05";
import Config from "./<%= context.packageConfigDirPrefix %>config.js";
//@@viewOff:imports

//@@viewOn:constants
//@@viewOff:constants

//@@viewOn:css
const Css = {
  main: () => Config.Css.css({}),
};
//@@viewOff:css

//@@viewOn:helpers
//@@viewOff:helpers

const <%= context.componentNameCamelCase %> = createVisualComponent({
  //@@viewOn:statics
  uu5Tag: Config.TAG + "<%= context.componentNameCamelCase %>",
  nestingLevel: ["areaCollection", "area"],
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {},
  //@@viewOff:propTypes

  //@@viewOn:defaultProps
  defaultProps: {},
  //@@viewOff:defaultProps

  render(props) {
    //@@viewOn:private
    const { children } = props;
    //@@viewOff:private

    //@@viewOn:interface
    //@@viewOff:interface

    //@@viewOn:render
    const attrs = Utils.VisualComponent.getAttrs(props, Css.main());
    const currentNestingLevel = Utils.NestingLevel.getNestingLevel(props, <%= context.componentNameCamelCase %>);

    return currentNestingLevel ? (
      <div {...attrs}>
        <div>Visual Component {<%= context.componentNameCamelCase %>.uu5Tag}</div>
        <Content nestingLevel={currentNestingLevel}>{children}</Content>
      </div>
    ) : null;
    //@@viewOff:render
  },
});

//@@viewOn:exports
export { <%= context.componentNameCamelCase %> };
export default <%= context.componentNameCamelCase %>;
//@@viewOff:exports
