// uu5tilesg02/dist/assets/mocks/animal-mocks.jsx
//@@viewOn:imports
import React from "react";
import { createVisualComponent, Utils, Lsi } from "uu5g05";
import { Config } from "uu5g05-dev";
import Uu5Elements from "uu5g05-elements";
import Uu5Forms from "uu5g05-forms";
import Uu5TilesElements from "uu5tilesg02-elements";
//@@viewOff:imports

const DATA = [
  {
    id: "tygr01",
    speciesName: "Bengal tiger",
    speciesTaxonomyName: "Panthera tigris",
    class: "Mammalia",
    order: "Carnivora",
    family: "Felidae",
    location: "Asia",
    description: "",
    img: "tygr_01",
  },
  {
    id: "pavian01",
    speciesName: "Hamadryas baboon",
    speciesTaxonomyName: "Papio hamadryas",
    class: "Mammalia",
    order: "Primates",
    family: "Cercopithecidae",
    location: "Africa",
    description: "",
    img: "pavian_01",
  },
  {
    id: "malpa01",
    speciesName: "Golden-bellied capuchin",
    speciesTaxonomyName: "Sapajus xanthostemos",
    class: "Mammalia",
    order: "Primates",
    family: "Cebidae",
    location: "America",
    description: "",
    img: "malpa_01",
  },
  {
    id: "langur01",
    speciesName: "Red-shanked douc",
    speciesTaxonomyName: "Pygathrix nemaeus",
    class: "Mammalia",
    order: "Primates",
    family: "Cercopithecidae",
    location: "Asia",
    description: "",
    img: "langur_01",
  },
  {
    id: "orangutan01",
    speciesName: "Sumatran orangutan",
    speciesTaxonomyName: "Pongo abelii",
    class: "Mammalia",
    order: "Primates",
    family: "Hominidae",
    location: "Asia",
    description: "",
    img: "orangutan_01",
  },
  {
    id: "varan02",
    speciesName: "Komodo dragon",
    speciesTaxonomyName: "Varanus komodoensis",
    class: "Reptilia",
    order: "Squamata",
    family: "Varanidae",
    location: "Asia",
    description: "",
    img: "varan_02",
  },
  {
    id: "krokodyl01",
    speciesName: "Saltwater crocodile",
    speciesTaxonomyName: "Crocodylus porosus",
    class: "Reptilia",
    order: "Crocodilia",
    family: "Crocodylidae",
    location: "Asia, Australia",
    description: "",
    img: "krokodyl_01",
  },
  {
    id: "mandril01",
    speciesName: "Mandrill",
    speciesTaxonomyName: "Mandrillus sphinx",
    class: "Mammalia",
    order: "Primates",
    family: "Cercopithecidae",
    location: "Africa",
    description: "",
    img: "mandril_01",
  },
  {
    id: "simpanz01",
    speciesName: "Common chimpanzee",
    speciesTaxonomyName: "Pan troglodytes",
    class: "Mammalia",
    order: "Primates",
    family: "Hominidae",
    location: "Africa",
    description: "",
    img: "simpanz_01",
  },
  {
    id: "kahau01",
    speciesName: "Proboscis monkey",
    speciesTaxonomyName: "Nasalis larvatus",
    class: "Mammalia",
    order: "Primates",
    family: "Cercopithecidae",
    location: "Asia",
    description: "",
    img: "kahau_01",
  },
  {
    id: "nosozec01",
    speciesName: "White rhinoceros",
    speciesTaxonomyName: "Ceratotherium simum",
    class: "Mammalia",
    order: "Perissodactyla",
    family: "Rhinocerotidae",
    location: "Africa",
    description: "",
    img: "nosorozec_01",
  },
  {
    id: "zelva02",
    speciesName: "Red-eared slider",
    speciesTaxonomyName: "Trachemys scripta",
    class: "Reptilia",
    order: "Testudines",
    family: "Emydidae",
    location: "America",
    description: "",
    img: "zelva_02",
  },
  {
    id: "skalnak01",
    speciesName: "Guianan cock-of-the-rock",
    speciesTaxonomyName: "Rupicola rupicola",
    class: "Aves",
    order: "Passeriformes",
    family: "Cotingidae",
    location: "America",
    description: "",
    img: "skalnak_01",
  },
  {
    id: "gavial01",
    speciesName: "Gharial",
    speciesTaxonomyName: "Gavialus gangeticus",
    class: "Reptilia",
    order: "Crocodilia",
    family: "Gavialidae",
    location: "Asia",
    description: "",
    img: "gavial_01",
  },
  {
    id: "zelva01",
    speciesName: "Southern river terrapin",
    speciesTaxonomyName: "Batagur affinis",
    class: "Reptilia",
    order: "Testudines",
    family: "Geoemydidae",
    location: "Asia",
    description: "",
    img: "zelva_01",
  },
  {
    id: "panda02",
    speciesName: "Red panda",
    speciesTaxonomyName: "Ailurus fulgens",
    class: "Mammalia",
    order: "Carnivora",
    family: "Ailuridae",
    location: "Asia",
    description: "",
    img: "panda_02",
  },
  {
    id: "panda01",
    speciesName: "Giant panda",
    speciesTaxonomyName: "Ailuropoda melanoleuca",
    class: "Mammalia",
    order: "Carnivora",
    family: "Ursidae",
    location: "Asia",
    description: "",
    img: "panda_01",
  },
  {
    id: "kapybara01",
    speciesName: "Capybara",
    speciesTaxonomyName: "Hydrochoreus hydrochaeris",
    class: "Mammalia",
    order: "Rodentia",
    family: "Caviidae",
    location: "America",
    description: "",
    img: "kapybara_01",
  },
  {
    id: "tukan01",
    speciesName: "Toco toucan",
    speciesTaxonomyName: "Ramphastos toco",
    class: "Aves",
    order: "Piciformes",
    family: "Ramphastidae",
    location: "America",
    description: "",
    img: "tukan_01",
  },
  {
    id: "aligator01",
    speciesName: "Chinese alligator",
    speciesTaxonomyName: "Alligator sinensis",
    class: "Reptilia",
    order: "Crocodilia",
    family: "Alligatoridae",
    location: "Asia",
    description: "",
    img: "aligator_01",
  },
  {
    id: "lachtan01",
    speciesName: "Brown fur seal",
    speciesTaxonomyName: "Arctocephalus pusillus",
    class: "Mammalia",
    order: "Carnivora",
    family: "Otariidae",
    location: "Africa",
    description: "",
    img: "lachtan_01",
  },
  {
    id: "plamenak01",
    speciesName: "Greater flamingo",
    speciesTaxonomyName: "Phoenicopterus roseus",
    class: "Aves",
    order: "Phoenicopteriformes",
    family: "Phoenicopteridae",
    location: "Europe, Asia, Africa",
    description: "",
    img: "plamenak_01",
  },
  {
    id: "chameleon01",
    speciesName: "Namaqua chameleon",
    speciesTaxonomyName: "Chameleo namaquensis",
    class: "Reptilia",
    order: "Squamata",
    family: "Chamaeleonidae",
    location: "Africa",
    description: "",
    img: "chameleon_01",
  },
  {
    id: "agama01",
    speciesName: "Agama agama",
    speciesTaxonomyName: "Agama agama",
    class: "Reptilia",
    order: "Squamata",
    family: "Agamidae",
    location: "Africa",
    description: "",
    img: "agama_01",
  },
  {
    id: "hyena01",
    speciesName: "Spotted hyena",
    speciesTaxonomyName: "Crocuta crocuta",
    class: "Mammalia",
    order: "Carnivora",
    family: "Hyaenidae",
    location: "Africa",
    description: "",
    img: "hyena_01",
  },
  {
    id: "sakal01",
    speciesName: "Black-backed jackal",
    speciesTaxonomyName: "Canis mesomelas",
    class: "Mammalia",
    order: "Carnivora",
    family: "Canidae",
    location: "Africa",
    description: "",
    img: "sakal_01",
  },
  {
    id: "pstros01",
    speciesName: "Common ostrich",
    speciesTaxonomyName: "Struthio camelus",
    class: "Aves",
    order: "Struthioniformes",
    family: "Struthionidae",
    location: "Africa",
    description: "",
    img: "pstros_01",
  },
  {
    id: "hadilov01",
    speciesName: "Secretarybird",
    speciesTaxonomyName: "Sagittarius serpentarius",
    class: "Aves",
    order: "Accipitriformes",
    family: "Sagittariidae",
    location: "Africa",
    description: "",
    img: "hadilov_01",
  },
  {
    id: "kudu01",
    speciesName: "Greater kudu",
    speciesTaxonomyName: "Tragelaphus strepsiceros",
    class: "Mammalia",
    order: "Artiodactyla",
    family: "Bovidae",
    location: "Africa",
    description: "",
    img: "kudu_01",
  },
  {
    id: "mamba01",
    speciesName: "Black mamba",
    speciesTaxonomyName: "Dendroaspis polylepis",
    class: "Reptilia",
    order: "Squamata",
    family: "Elapidae",
    location: "Africa",
    description: "",
    img: "mamba_01",
  },
  {
    id: "zebra01",
    speciesName: "Burchell's zebra",
    speciesTaxonomyName: "Equus quagga burchellii",
    class: "Mammalia",
    order: "Perissodactyla",
    family: "Equidae",
    location: "Africa",
    description: "",
    img: "zebra_01",
  },
  {
    id: "dikdik01",
    speciesName: "Kirk's dik-dik",
    speciesTaxonomyName: "Madoqua kirkii",
    class: "Mammalia",
    order: "Artiodactyla",
    family: "Bovidae",
    location: "Africa",
    description: "",
    img: "dikdik_01",
  },
  {
    id: "pakun01",
    speciesName: "Blue wildebeest",
    speciesTaxonomyName: "Connochaetes taurinus",
    class: "Mammalia",
    order: "Artiodactyla",
    family: "Bovidae",
    location: "Africa",
    description: "",
    img: "pakun_01",
  },
  {
    id: "buvolec01",
    speciesName: "Common tsessebe",
    speciesTaxonomyName: "Damaliscus lunatus",
    class: "Mammalia",
    order: "Artiodactyla",
    family: "Bovidae",
    location: "Africa",
    description: "",
    img: "buvolec_01",
  },
  {
    id: "impala01",
    speciesName: "Impala",
    speciesTaxonomyName: "Aepyceros melampus",
    class: "Mammalia",
    order: "Artiodactyla",
    family: "Bovidae",
    location: "Africa",
    description: "",
    img: "impala_01",
  },
  {
    id: "vazka02",
    speciesName: "Yellow-winged darter",
    speciesTaxonomyName: "Trithemis kirbyi",
    class: "Insecta",
    order: "Odonata",
    family: "Libellulidae",
    location: "Africa",
    description: "",
    img: "vazka_02",
  },
  {
    id: "voduska01",
    speciesName: "Lechwe",
    speciesTaxonomyName: "Kobus leche",
    class: "Mammalia",
    order: "Artiodactyla",
    family: "Bovidae",
    location: "Africa",
    description: "",
    img: "voduska_01",
  },
  {
    id: "zoborozec01",
    speciesName: "Southern yellow-billed hornbill",
    speciesTaxonomyName: "Tckus leucomelas",
    class: "Aves",
    order: "Bucerotiformes",
    family: "Bucerotidae",
    location: "Africa",
    description: "",
    img: "zoborozec_01",
  },
  {
    id: "vlha01",
    speciesName: "White-fronted bee-eater",
    speciesTaxonomyName: "Merops bullockoides",
    class: "Aves",
    order: "Coraciiformes",
    family: "Meropidae",
    location: "Africa",
    description: "",
    img: "vlha_01",
  },
  {
    id: "lednacek01",
    speciesName: "Pied kingfisher",
    speciesTaxonomyName: "Ceryle rudis",
    class: "Aves",
    order: "Coraciiformes",
    family: "Alcedinidae",
    location: "Africa",
    description: "",
    img: "lednacek_01",
  },
  {
    id: "varan01",
    speciesName: "Nile monitor",
    speciesTaxonomyName: "Varanus niloticus",
    class: "Reptilia",
    order: "Squamata",
    family: "Varanidae",
    location: "Africa",
    description: "",
    img: "varan_01",
  },
  {
    id: "slon01",
    speciesName: "African bush elephant",
    speciesTaxonomyName: "Loxodonta africana",
    class: "Mammalia",
    order: "Proboscidea",
    family: "Elephantidae",
    location: "Africa",
    description: "",
    img: "slon_01",
  },
  {
    id: "nesyt01",
    speciesName: "Yellow-billed stork",
    speciesTaxonomyName: "Mycteria ibis",
    class: "Aves",
    order: "Ciconiiformes",
    family: "Ciconiidae",
    location: "Africa",
    description: "",
    img: "nesyt_01",
  },
  {
    id: "perlicka01",
    speciesName: "Helmeted guineafowl",
    speciesTaxonomyName: "Numida meleagris",
    class: "Aves",
    order: "Galliformes",
    family: "Numididae",
    location: "Africa",
    description: "",
    img: "perlicka_01",
  },
  {
    id: "buvol01",
    speciesName: "African buffalo",
    speciesTaxonomyName: "Syncerus cafer",
    class: "Mammalia",
    order: "Artiodactyla",
    family: "Bovidae",
    location: "Africa",
    description: "",
    img: "buvol_01",
  },
  {
    id: "mandelik01",
    speciesName: "Lilac-breasted roller",
    speciesTaxonomyName: "Coracias caudatus",
    class: "Aves",
    order: "Coraciiformes",
    family: "Coraciidae",
    location: "Africa",
    description: "",
    img: "mandelik_01",
  },
  {
    id: "hroch01",
    speciesName: "Hippopotamus",
    speciesTaxonomyName: "Hippopotamus amphibius",
    class: "Mammalia",
    order: "Artiodactyla",
    family: "Hippopotamidae",
    location: "Africa",
    description: "",
    img: "hroch_01",
  },
  {
    id: "kockodan01",
    speciesName: "Green monkey",
    speciesTaxonomyName: "Chlorocebus sabaeus",
    class: "Mammalia",
    order: "Primates",
    family: "Cercopithecidae",
    location: "Africa",
    description: "",
    img: "kockodan_01",
  },
  {
    id: "prase01",
    speciesName: "Desert warthog",
    speciesTaxonomyName: "Phacochoerus aethiopicus",
    class: "Mammalia",
    order: "Artiodactyla",
    family: "Suidae",
    location: "Africa",
    description: "",
    img: "prase_01",
  },
  {
    id: "slon02",
    speciesName: "African bush elephant",
    speciesTaxonomyName: "Loxodonta africana",
    class: "Mammalia",
    order: "Proboscidea",
    family: "Elephantidae",
    location: "Africa",
    description: "",
    img: "slon_02",
  },
  {
    id: "sup01",
    speciesName: "Cape vulture",
    speciesTaxonomyName: "Gyps coprotheres",
    class: "Aves",
    order: "Accipitriformes",
    family: "Accipitridae",
    location: "Africa",
    description: "",
    img: "sup_01",
  },
  {
    id: "lev01",
    speciesName: "Lion",
    speciesTaxonomyName: "Panthera leo",
    class: "Mammalia",
    order: "Carnivora",
    family: "Felidae",
    location: "Africa, Asia",
    description: "",
    img: "lev_01",
  },
  {
    id: "krab01",
    speciesName: "Round-eyed ghost crab",
    speciesTaxonomyName: "Ocypode cordimana",
    class: "Malacostraca",
    order: "Decapoda",
    family: "Ocypodidae",
    location: "Africa",
    description: "",
    img: "krab_01",
  },
  {
    id: "kalon01",
    speciesName: "Seychelles fruit bat",
    speciesTaxonomyName: "Pteropus seychellensis",
    class: "Mammalia",
    order: "Chiroptera",
    family: "Pteropodidae",
    location: "Africa",
    description: "",
    img: "kalon_01",
  },
  {
    id: "ryba02",
    speciesName: "Koran angelfish",
    speciesTaxonomyName: "Pomacanthus semicirculatus",
    class: "Actinopterygii",
    order: "Perciformes",
    family: "Pomacanthidae",
    location: "Indian Ocean",
    description: "",
    img: "ryba_02",
  },
  {
    id: "ryba01",
    speciesName: "Ornate wrasse",
    speciesTaxonomyName: "Thalassoma pavo",
    class: "Actinopterygii",
    order: "Perciformes",
    family: "Labridae",
    location: "Indian Ocean",
    description: "",
    img: "ryba_01",
  },
  {
    id: "rejnok01",
    speciesName: "Porcupine ray",
    speciesTaxonomyName: "Urogymnus asperrimus",
    class: "Chondrichthyes",
    order: "Myliobatiformes",
    family: "Dasyatidae",
    location: "Indian Ocean",
    description: "",
    img: "rejnok_01",
  },
  {
    id: "strdimil01",
    speciesName: "Seychelles sunbird",
    speciesTaxonomyName: "Cinnyris dussumieri",
    class: "Aves",
    order: "Passeriformes",
    family: "Nectariniidae",
    location: "Africa",
    description: "",
    img: "strdimil_01",
  },
  {
    id: "kareta01",
    speciesName: "Hawksbill sea turtle",
    speciesTaxonomyName: "Eretmochelys imbricata",
    class: "Reptilia",
    order: "Testudines",
    family: "Cheloniidae",
    location: "Oceans",
    description: "",
    img: "kareta_01",
  },
  {
    id: "krab02",
    speciesName: "Cardisoma carnifex",
    speciesTaxonomyName: "Cardisoma carnifex",
    class: "Malacostraca",
    order: "Decapoda",
    family: "Gecarcinidae",
    location: "Africa, Asia",
    description: "",
    img: "krab_02",
  },
  {
    id: "vazka01",
    speciesName: "Four-spotted chaser",
    speciesTaxonomyName: "Libellula quadrimaculata",
    class: "Insecta",
    order: "Odonata",
    family: "Libellulidae",
    location: "Europe",
    description: "",
    img: "vazka_01",
  },
  {
    id: "skokan01",
    speciesName: "Marsh frog",
    speciesTaxonomyName: "Pelophylax ridibundus",
    class: "Amphibia",
    order: "Anura",
    family: "Ranidae",
    location: "Europe, Asia",
    description: "",
    img: "skokan_01",
  },
  {
    id: "kolpik01",
    speciesName: "Eurasian spoonbill",
    speciesTaxonomyName: "Platalea leucorodia",
    class: "Aves",
    order: "Pelecaniformes",
    family: "Threskiornithidae",
    location: "Europe, Asia, Africa",
    description: "",
    img: "kolpik_01",
  },
  {
    id: "nosal01",
    speciesName: "White-nosed coati",
    speciesTaxonomyName: "Nasua narica",
    class: "Mammalia",
    order: "Carnivora",
    family: "Procyonidae",
    location: "America",
    description: "",
    img: "nosal_01",
  },
  {
    id: "tapir01",
    speciesName: "Malayan tapir",
    speciesTaxonomyName: "Tapirus indicus",
    class: "Mammalia",
    order: "Perissodactyla",
    family: "Tapiridae",
    location: "Asia",
    description: "",
    img: "tapir_01",
  },
  {
    id: "volavka01",
    speciesName: "Grey heron",
    speciesTaxonomyName: "Ardea cinerea",
    class: "Aves",
    order: "Pelecaniformes",
    family: "Ardeidae",
    location: "Europe, Asia, Africa",
    description: "",
    img: "volavka_01",
  },
  {
    id: "motyl02",
    speciesName: "Ambrax butterfly",
    speciesTaxonomyName: "Papilio ambrax",
    class: "Insecta",
    order: "Lepidoptera",
    family: "Papilionidae",
    location: "Australia",
    description: "",
    img: "motyl_02",
  },
  {
    id: "motyl01",
    speciesName: "Paper kite",
    speciesTaxonomyName: "Idea leuconoe",
    class: "Insecta",
    order: "Lepidoptera",
    family: "Nymphalidae",
    location: "Asia, Australia",
    description: "",
    img: "motyl_01",
  },
  {
    id: "lvicek01",
    speciesName: "Golden-headed lion tamarin",
    speciesTaxonomyName: "Leontopithecus chrysomelas",
    class: "Mammalia",
    order: "Primates",
    family: "Callitrichidae",
    location: "America",
    description: "",
    img: "lvicek_01",
  },
  {
    id: "pralesnicka01",
    speciesName: "Yellow-banded poison dart frog",
    speciesTaxonomyName: "Dendrobates leucomelas",
    class: "Amphibia",
    order: "Anura",
    family: "Dendrobatidae",
    location: "America",
    description: "",
    img: "pralesnicka_01",
  },
  {
    id: "pirana01",
    speciesName: "Red-bellied piranha",
    speciesTaxonomyName: "Pygocentrus nattereri",
    class: "Actinopterygii",
    order: "Characiformes",
    family: "Serrasalmidae",
    location: "America",
    description: "",
    img: "pirana_01",
  },
  {
    id: "klaun01",
    speciesName: "Red Sea clownfish",
    speciesTaxonomyName: "Amphiprion bicinctus",
    class: "Actinopterygii",
    order: "Perciformes",
    family: "Pomacentridae",
    location: "Indian Ocean",
    description: "",
    img: "klaun_01",
  },
  {
    id: "kozoroh01",
    speciesName: "Alpine ibex",
    speciesTaxonomyName: "Capra ibex",
    class: "Mammalia",
    order: "Artiodactyla",
    family: "Bovidae",
    location: "Europe",
    description: "",
    img: "kozoroh_01",
  },
  {
    id: "irbis01",
    speciesName: "Snow leopard",
    speciesTaxonomyName: "Panthera uncia",
    class: "Mammalia",
    order: "Carnivora",
    family: "Felidae",
    location: "Asia",
    description: "",
    img: "irbis_01",
  },
  {
    id: "sovice01",
    speciesName: "Snowy owl",
    speciesTaxonomyName: "Bubo scandiacus",
    class: "Aves",
    order: "Strigiformes",
    family: "Strigidae",
    location: "Europe, Asia, America",
    description: "",
    img: "sovice_01",
  },
];

const CLASS_LIST = [];
for (let item of DATA) {
  let itemClass = item.class;
  if (CLASS_LIST.indexOf(itemClass) === -1) CLASS_LIST.push(itemClass);
}

const ORDER_LIST = [];
for (let item of DATA) {
  let itemOrder = item.order;
  if (ORDER_LIST.indexOf(itemOrder) === -1) ORDER_LIST.push(itemOrder);
}

const FAMILY_LIST = [];
for (let item of DATA) {
  let itemFamily = item.family;
  if (FAMILY_LIST.indexOf(itemFamily) === -1) FAMILY_LIST.push(itemFamily);
}

const LOCATION_LIST = [];
for (let item of DATA) {
  let itemLocation = item.location;
  let locationList = itemLocation.split(",");
  for (let locationItem of locationList) {
    locationItem = locationItem.trim();
    if (LOCATION_LIST.indexOf(locationItem) === -1) LOCATION_LIST.push(locationItem);
  }
}

const IMG_LABEL = { cs: "Obrázek", en: "Image" };
const NAME_LABEL = { cs: "Název", en: "Species name" };
const TAXONOMY_NAME_LABEL = { cs: "Latinský název", en: "Latin name" };
const CLASS_LABEL = { cs: "Třída", en: "Class" };
const ORDER_LABEL = { cs: "Řád", en: "Order" };
const FAMILY_LABEL = { cs: "Čeleď", en: "Family" };
const LOCATION_LABEL = { cs: "Výskyt", en: "Location" };
const DESCRIPTION_LABEL = { cs: "Popis", en: "Description" };

const SORTER_LIST = [
  {
    key: "speciesName",
    label: NAME_LABEL,
    sort: (a, b) => a.speciesName.localeCompare(b.speciesName),
  },
  {
    key: "speciesTaxonomyName",
    label: TAXONOMY_NAME_LABEL,
    sort: (a, b) => a.speciesTaxonomyName.localeCompare(b.speciesTaxonomyName),
  },
  {
    key: "class",
    label: CLASS_LABEL,
    sort: (a, b) => a.class.localeCompare(b.class),
  },
  {
    key: "order",
    label: ORDER_LABEL,
    sort: (a, b) => a.order.localeCompare(b.order),
  },
  {
    key: "family",
    label: FAMILY_LABEL,
    sort: (a, b) => a.family.localeCompare(b.family),
  },
  {
    key: "location",
    label: LOCATION_LABEL,
    sort: (a, b) => a.location.localeCompare(b.location),
  },
];

const FILTER_LIST = [
  {
    key: "speciesName",
    label: NAME_LABEL,
    filter: (item, value) => {
      let fragments = value.split(/[\s,.-;:_]/);
      return fragments.some((frag) => {
        let itemValue =
          typeof item.speciesName === "object" ? Utils.Language.getItem(item.speciesName) : item.speciesName;
        return itemValue.toLowerCase().indexOf(frag.toLowerCase()) !== -1;
      });
    },
    inputProps: { placeholder: { en: "Enter value", cs: "Zadejte hodnotu" } },
  },
  {
    key: "speciesTaxonomyName",
    label: TAXONOMY_NAME_LABEL,
    filter: (item, value) => {
      let fragments = value.split(/[\s,.-;:_]/);
      return fragments.some((frag) => item.speciesTaxonomyName.toLowerCase().indexOf(frag.toLowerCase()) !== -1);
    },
    inputProps: { placeholder: { en: "Enter value", cs: "Zadejte hodnotu" } },
  },
  {
    key: "location",
    label: LOCATION_LABEL,
    filter: (item, value) => {
      return value.some((frag) => {
        let itemValue = typeof item.location === "object" ? Utils.Language.getItem(item.location) : item.location;
        return itemValue.toLowerCase().indexOf(frag.toLowerCase()) !== -1;
      });
    },
    inputType: "text-select",
    inputProps: {
      multiple: true,
      itemList: LOCATION_LIST.map((it) => ({ value: it, children: it })),
      placeholder: { en: "Enter value", cs: "Zadejte hodnotu" },
    },
  },
  {
    key: "class",
    label: CLASS_LABEL,
    filter: (item, value) => {
      return value.some((frag) => {
        let itemValue = typeof item.class === "object" ? Utils.Language.getItem(item.class) : item.class;
        return itemValue.toLowerCase().indexOf(frag.toLowerCase()) !== -1;
      });
    },
    inputType: "select",
    inputProps: {
      multiple: true,
      itemList: CLASS_LIST.map((it) => ({ value: it, children: it })),
      placeholder: { en: "Enter value", cs: "Zadejte hodnotu" },
    },
  },
  {
    key: "order",
    label: ORDER_LABEL,
    filter: (item, value) => {
      let itemValue = typeof item.order === "object" ? Utils.Language.getItem(item.order) : item.order;
      return itemValue.toLowerCase().indexOf(value.toLowerCase()) !== -1;
    },
    inputType: "text-select",
    inputProps: {
      itemList: ORDER_LIST.map((it) => ({ value: it, children: it })),
      placeholder: { en: "Enter value", cs: "Zadejte hodnotu" },
    },
  },
  {
    key: "family",
    label: FAMILY_LABEL,
    filter: (item, value) => {
      let itemValue = typeof item.family === "object" ? Utils.Language.getItem(item.family) : item.family;
      return itemValue.toLowerCase().indexOf(value.toLowerCase()) !== -1;
    },
    inputType: "select",
    inputProps: {
      itemList: FAMILY_LIST.map((it) => ({ value: it, children: it })),
      placeholder: { en: "Enter value", cs: "Zadejte hodnotu" },
    },
  },
];

const ImageRenderer = createVisualComponent({
  uu5Tag: Config.TAG + "ImageRenderer",
  render(props) {
    const { data } = props;
    return (
      <div className={Config.Css.css({ minWidth: 36, display: "inline-block" })}>
        <img
          src={`https://uuapp.plus4u.net/uu-webkit-maing02/1bdb540a0c53458aa20cfd9c179e6a1c/${data.img}`}
          className={Config.Css.css({ height: 24, display: "inline-block" })}
        />
      </div>
    );
  },
});

const SpeciesNameRenderer = createVisualComponent({
  uu5Tag: Config.TAG + "SpeciesNameRenderer",
  render(props) {
    const { data } = props;
    return data.speciesName;
  },
});

const SpeciesTaxonomyNameRenderer = createVisualComponent({
  uu5Tag: Config.TAG + "SpeciesTaxonomyNameRenderer",
  render(props) {
    const { data } = props;
    return data.speciesTaxonomyName;
  },
});

const ClassRenderer = createVisualComponent({
  uu5Tag: Config.TAG + "ClassRenderer",
  render(props) {
    const { data } = props;
    return data.class;
  },
});

const OrderRenderer = createVisualComponent({
  uu5Tag: Config.TAG + "OrderRenderer",
  render(props) {
    const { data } = props;
    return data.order;
  },
});

const FamilyRenderer = createVisualComponent({
  uu5Tag: Config.TAG + "FamilyRenderer",
  render(props) {
    const { data } = props;
    return data.family;
  },
});

const LocationRenderer = createVisualComponent({
  uu5Tag: Config.TAG + "LocationRenderer",
  render(props) {
    const { data } = props;
    return data.location;
  },
});

const DescriptionRenderer = createVisualComponent({
  uu5Tag: Config.TAG + "DescriptionRenderer",
  render(props) {
    const { data } = props;
    return data.description;
  },
});

const SERIE_LIST = [
  { value: "img", label: IMG_LABEL, dataItem: ImageRenderer },
  { value: "speciesName", label: NAME_LABEL, dataItem: SpeciesNameRenderer, visible: "always" },
  { value: "speciesTaxonomyName", label: TAXONOMY_NAME_LABEL, dataItem: SpeciesTaxonomyNameRenderer },
  { value: "class", label: CLASS_LABEL, dataItem: ClassRenderer },
  { value: "order", label: ORDER_LABEL, dataItem: OrderRenderer },
  { value: "family", label: FAMILY_LABEL, dataItem: FamilyRenderer },
  { value: "location", label: LOCATION_LABEL, dataItem: LocationRenderer },
  { value: "description", label: DESCRIPTION_LABEL, dataItem: DescriptionRenderer, visible: false, fixed: "end" },
];

const COLUMN_LIST = [
  {
    value: "img",
    header: () => {}, // To create an empty header cell
    cell: ImageRenderer,
    cellComponent: <Uu5TilesElements.Table.Cell horizontalAlignment="center" verticalAlignment="center" />,
    minWidth: 64,
    maxWidth: "auto",
  },
  { value: "speciesName", header: <Lsi lsi={NAME_LABEL} />, minWidth: 250, maxWidth: 350 },
  { value: "speciesTaxonomyName", header: <Lsi lsi={TAXONOMY_NAME_LABEL} />, minWidth: 100 },
  { value: "class", header: <Lsi lsi={CLASS_LABEL} />, minWidth: 130 },
  { value: "order", header: <Lsi lsi={ORDER_LABEL} />, minWidth: 150 },
  { value: "family", header: <Lsi lsi={FAMILY_LABEL} />, minWidth: 150 },
  { value: "location", header: <Lsi lsi={LOCATION_LABEL} />, minWidth: 150 },
  { value: "description", header: <Lsi lsi={DESCRIPTION_LABEL} />, minWidth: 100 },
];

const PRESET_LIST = [
  {
    value: "africanAnimals",
    label: { cs: "Africká zvířata", en: "African animals" },
    filterList: [{ key: "location", value: ["Africa"] }],
    serieList: SERIE_LIST,
  },
  {
    value: "africanMammals",
    label: { cs: "Afričtí savci (A-Z)", en: "African mammals (A-Z)" },
    filterList: [
      { key: "location", value: ["Africa"] },
      { key: "class", value: ["Mammalia"] },
    ],
    sorterList: [{ key: "speciesName", ascending: true }],
    serieList: SERIE_LIST,
  },
];

const VIEW_LIST = [
  { label: { en: "Table", cs: "Tabulka" }, icon: "uugds-view-list", value: "table" },
  { label: { en: "Grid", cs: "Grid" }, icon: "uugds-view-grid", value: "grid" },
];

const Tile = createVisualComponent({
  //@@viewOn:statics
  uu5Tag: "Uu5TilesElements.Mock.Tile",
  //@@viewOff:statics

  //@@viewOn:propTypes
  propTypes: {},
  //@@viewOff:propTypes

  //@@viewOn:defaultProps
  defaultProps: {},
  //@@viewOff:defaultProps

  render(props) {
    //@@viewOn:private
    let { data, ...otherProps } = props;
    //@@viewOff:private

    //@@viewOn:interface
    //@@viewOff:interface

    //@@viewOn:render
    return (
      <Uu5TilesElements.Tile {...otherProps} headerOverlap>
        {({ padding }) => {
          return (
            <>
              <img
                src={`https://uuapp.plus4u.net/uu-webkit-maing02/1bdb540a0c53458aa20cfd9c179e6a1c/${data.img}`}
                alt={data.img}
                className={Config.Css.css({ width: "100%", height: 120, display: "block", objectFit: "cover" })}
              />
              <div
                className={Config.Css.css({
                  paddingTop: padding.top,
                  paddingRight: padding.right,
                  paddingBottom: padding.bottom,
                  paddingLeft: padding.left,
                })}
              >
                <div>
                  <Uu5Elements.Icon icon="uugds-favorites" />
                  <strong>{data.speciesName}</strong>
                  <em>
                    <small>({data.speciesTaxonomyName})</small>
                  </em>
                </div>
                <Uu5Elements.Text key="location" colorScheme="green">
                  <Uu5Elements.Icon icon="uugdsstencil-navigation-mapmarker-solid" />
                  {data.location}
                </Uu5Elements.Text>
                <div key="description"></div>
                <small key="taxonomy">
                  <div>
                    Class{"\xA0"}
                    <strong>{data.class}</strong>, Order{"\xA0"}
                    <strong>{data.order}</strong>, Family{"\xA0"}
                    <strong>{data.family}</strong>
                  </div>
                </small>
              </div>
            </>
          );
        }}
      </Uu5TilesElements.Tile>
    );
    //@@viewOff:render
  },
});

export { DATA, SORTER_LIST, FILTER_LIST, COLUMN_LIST, SERIE_LIST, PRESET_LIST, VIEW_LIST, Tile };
