// uu5tilesg02/dist/assets/mocks/learning-data-mocks.jsx
const DATA = [
  {
    id: "tiger01",
    name: "Bengal tiger",
    class: "Mammalia",
    order: "Carnivora",
    family: "Felidae",
    location: "Asia",
  },
  {
    id: "baboon02",
    name: "Hamadryas baboon",
    class: "Mammalia",
    order: "Primates",
    family: "Cercopithecidae",
    location: "Africa",
  },
  {
    id: "capuchin03",
    name: "Golden-bellied capuchin",
    class: "Mammalia",
    family: "Cebidae",
    location: "America",
  },
  {
    id: "ibex04",
    name: "Alpine ibex",
    class: "Mammalia",
    order: "Artiodactyla",
    family: "Bovidae",
    location: "Europe",
  },
  {
    id: "clownfish05",
    name: "Red Sea clownfish",
    class: "Actinopterygii",
    order: "Perciformes",
    location: "Indian Ocean",
  },
  {
    id: "dragonfly06",
    name: "Four-spotted chaser",
    class: "Insecta",
    order: "Odonata",
    family: "Libellulidae",
    location: "Europe",
  },
  {
    id: "turtle07",
    name: "Hawksbill sea turtle",
    class: "Reptilia",
    order: "Testudines",
    family: "Cheloniidae",
    location: "Oceans",
  },
  {
    id: "monkey08",
    name: "Red-shanked douc",
    class: "Mammalia",
    order: "Primates",
    family: "Cercopithecidae",
    location: "Asia",
  },
];

export { DATA };
